From 4d31f75978705755784848cbd64d36ae5dbe4b85 Mon Sep 17 00:00:00 2001
From: Nirmal Raj S <nirm@qti.qualcomm.com>
Date: Mon, 2 Mar 2026 10:35:53 +0530
Subject: [PATCH] updated ieee80211w value as 1 for sae-mixed mode

[NOTE] 1. skip adding notify_mgmt_frames for BH AP Vap
       2. added link active check before trigger RELOAD in refresh_radio_info
       3. added zero value check to skip new config addition

Signed-off-by: Nirmal Raj S <nirm@qti.qualcomm.com>
---
 .../backhaul_manager/backhaul_manager.cpp     | 43 +++++++++++++++++++
 .../bcl/include/bcl/beerocks_os_utils.h       | 10 ++++-
 .../beerocks/bcl/source/beerocks_os_utils.cpp | 16 +++++++
 .../bwl/nl80211/ap_wlan_hal_nl80211.cpp       | 36 ++++++++++++----
 .../beerocks/hostapd/source/configuration.cpp |  2 +-
 5 files changed, 96 insertions(+), 11 deletions(-)

diff --git a/agent/src/beerocks/slave/backhaul_manager/backhaul_manager.cpp b/agent/src/beerocks/slave/backhaul_manager/backhaul_manager.cpp
index 3050ff5e3..eb03ec726 100644
--- a/agent/src/beerocks/slave/backhaul_manager/backhaul_manager.cpp
+++ b/agent/src/beerocks/slave/backhaul_manager/backhaul_manager.cpp
@@ -18,6 +18,7 @@
 #include "../tasks/topology_task.h"
 #include <bcl/beerocks_cmdu_client_factory_factory.h>
 #include <bcl/beerocks_cmdu_server_factory.h>
+#include <bcl/beerocks_os_utils.h>
 #include <bcl/beerocks_timer_factory_impl.h>
 #include <bcl/beerocks_timer_manager_impl.h>
 #include <bcl/beerocks_ucc_server_factory.h>
@@ -932,6 +933,27 @@ bool BackhaulManager::backhaul_fsm_main(bool &skip_select)
         m_task_pool.send_event(eTaskType::TOPOLOGY, TopologyTask::eEvent::AGENT_DEVICE_INITIALIZED);
 
         m_backhaul_switch = false;
+
+        // Print backhaul type to console before moving to OPERATIONAL state
+        // Use system() to echo to console since stdout is redirected to log file
+        if (!db->device_conf.local_gw) {
+            std::string timestamp = beerocks::os_utils::get_timestamp();
+            if (db->backhaul.connection_type == AgentDB::sBackhaul::eConnectionType::Wired) {
+                std::string msg = "echo '" + timestamp + "Backhaul Type: Wired' > /dev/console";
+                beerocks::os_utils::system_call(msg);
+                LOG(INFO) << timestamp << "Backhaul Type: Wired";
+            } else if (db->backhaul.connection_type ==
+                       AgentDB::sBackhaul::eConnectionType::Wireless) {
+                std::string msg = "echo '" + timestamp + "Backhaul Type: Wireless' > /dev/console";
+                beerocks::os_utils::system_call(msg);
+                LOG(INFO) << timestamp << "Backhaul Type: Wireless";
+            } else {
+                std::string msg =
+                    "echo '" + timestamp + "Backhaul Type: Invalid/Unknown' > /dev/console";
+                beerocks::os_utils::system_call(msg);
+                LOG(INFO) << timestamp << "Backhaul Type: Invalid/Unknown";
+            }
+        }
         FSM_MOVE_STATE(OPERATIONAL);
         break;
     }
@@ -971,6 +993,12 @@ bool BackhaulManager::backhaul_fsm_main(bool &skip_select)
             bool eth_link_up = beerocks::net::network_utils::linux_iface_is_up_and_running(db->ethernet.wan.iface_name);
             if (eth_link_up != m_eth_link_up) {
                 //update the previous bh connection
+                std::string timestamp = beerocks::os_utils::get_timestamp();
+                std::string msg =
+                    "echo '" + timestamp + "Backhaul state change detected:' > /dev/console";
+                beerocks::os_utils::system_call(msg);
+                LOG(INFO) << timestamp << "Backhaul state change detected:";
+
                 if (m_eth_link_up) {
                     m_prev_bh_connection = backhaul_connection::Wired;
                 } else {
@@ -1046,6 +1074,21 @@ bool BackhaulManager::backhaul_fsm_main(bool &skip_select)
                         break;
                     }
                 }
+
+                // Print backhaul switching message to console
+                std::string timestamp2 = beerocks::os_utils::get_timestamp();
+                if (m_eth_link_up) {
+                    std::string msg = "echo '" + timestamp2 +
+                                      "Backhaul Switching: Wireless -> Wired' > /dev/console";
+                    beerocks::os_utils::system_call(msg);
+                    LOG(INFO) << timestamp2 << "Backhaul Switching: Wireless -> Wired";
+                } else {
+                    std::string msg = "echo '" + timestamp2 +
+                                      "Backhaul Switching: Wired -> Wireless' > /dev/console";
+                    beerocks::os_utils::system_call(msg);
+                    LOG(INFO) << timestamp2 << "Backhaul Switching: Wired -> Wireless";
+                }
+
                 FSM_MOVE_STATE(RESTART);
             }
         }
diff --git a/common/beerocks/bcl/include/bcl/beerocks_os_utils.h b/common/beerocks/bcl/include/bcl/beerocks_os_utils.h
index 98f525001..11ec574d6 100644
--- a/common/beerocks/bcl/include/bcl/beerocks_os_utils.h
+++ b/common/beerocks/bcl/include/bcl/beerocks_os_utils.h
@@ -125,13 +125,21 @@ public:
      * @return std::string containing operating system name and version. 
      */
     static std::string get_os_name();
-    
+
     ///
     /// @brief Function that checks if wpa ctrl socket exists
     ///
     /// @return boolean that is true if the socket exists and false otherwise.
     ///
     static bool wpa_ctrl_skt_exists(const std::string &path);
+
+    /**
+     * @brief Get current timestamp string with uptime in seconds and microseconds.
+     *
+     * @return std::string containing formatted timestamp in the format "[ ssss.uuuuuu ] "
+     *         where ssss is seconds since epoch and uuuuuu is microseconds.
+     */
+    static std::string get_timestamp();
 };
 } // namespace beerocks
 
diff --git a/common/beerocks/bcl/source/beerocks_os_utils.cpp b/common/beerocks/bcl/source/beerocks_os_utils.cpp
index 867cc5769..e8078c289 100644
--- a/common/beerocks/bcl/source/beerocks_os_utils.cpp
+++ b/common/beerocks/bcl/source/beerocks_os_utils.cpp
@@ -12,7 +12,9 @@
 #include <array>
 #include <chrono>
 #include <fcntl.h>
+#include <iomanip>
 #include <linux/limits.h>
+#include <sstream>
 #include <sys/stat.h>
 #include <sys/types.h>
 #include <sys/utsname.h>
@@ -262,3 +264,17 @@ bool os_utils::wpa_ctrl_skt_exists(const std::string &path)
     struct stat st;
     return ((stat(path.c_str(), &st) == 0) && S_ISSOCK(st.st_mode));
 }
+
+std::string os_utils::get_timestamp()
+{
+    std::chrono::steady_clock::time_point now = std::chrono::steady_clock::now();
+    auto uptime_duration                      = now.time_since_epoch();
+    auto seconds = std::chrono::duration_cast<std::chrono::seconds>(uptime_duration);
+    auto microseconds =
+        std::chrono::duration_cast<std::chrono::microseconds>(uptime_duration) % 1000000;
+
+    std::stringstream ss;
+    ss << "[ " << std::setw(4) << std::setfill(' ') << seconds.count() << "." << std::setw(6)
+       << std::setfill('0') << microseconds.count() << " ] ";
+    return ss.str();
+}
diff --git a/common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp b/common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp
index 195d982b9..d75c6d666 100644
--- a/common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp
+++ b/common/beerocks/bwl/nl80211/ap_wlan_hal_nl80211.cpp
@@ -654,7 +654,7 @@ static bool assign_auth_encr_parameters(prplmesh::hostapd::Configuration &conf,
         }
         wpa_passphrase.assign(bss.network_key);
 
-        ieee80211w.assign("2");
+        ieee80211w.assign("1");
         disable_pmksa_caching.assign("0");
         okc.assign("1");
         wpa_disable_eapol_key_retries.assign("0");
@@ -757,12 +757,27 @@ bool ap_wlan_hal_nl80211::refresh_radio_info()
             }
         } else if (base_interface != m_radio_info.iface_name) {
             LOG(DEBUG) << "Base Iface " << m_radio_info.iface_name << " is a link";
-            // Check if hostapd control socket exists for m_radio_info.iface_name
-            std::string hostapd_ctrl_path = "/var/run/hostapd/" + m_radio_info.iface_name;
-            if (beerocks::os_utils::file_exists(hostapd_ctrl_path)) {
-                LOG(INFO) << "Sending RELOAD to " << m_radio_info.iface_name;
-                if (!wpa_ctrl_send_msg("RELOAD", m_radio_info.iface_name)) {
-                    LOG(ERROR) << "RELOAD command to hostapd failed for " << m_radio_info.iface_name;
+
+            // Extract link_id from the interface name
+            int link_id = -1;
+            extractLinkId(m_radio_info.iface_name, link_id);
+
+            // Check if the link is operational before sending RELOAD
+            if (link_id >= 0) {
+                bool link_operational = m_nl80211_client->check_link_id_operational(m_radio_info.iface_name, link_id);
+                LOG(DEBUG) << "Link " << link_id << " operational status: " << (link_operational ? "true" : "false");
+
+                if (!link_operational) {
+                    LOG(WARNING) << "Link " << link_id << " is not operational, trigger RELOAD";
+                    std::string hostapd_ctrl_path = "/var/run/hostapd/" + m_radio_info.iface_name;
+                    if (beerocks::os_utils::file_exists(hostapd_ctrl_path)) {
+                        LOG(INFO) << "Sending RELOAD to " << m_radio_info.iface_name;
+                        if (!wpa_ctrl_send_msg("RELOAD", m_radio_info.iface_name)) {
+                            LOG(ERROR) << "RELOAD command to hostapd failed for " << m_radio_info.iface_name;
+                        }
+                    }
+                } else {
+                    LOG(DEBUG) << "iface " <<  m_radio_info.iface_name << " is in up state, skip reload";
                 }
             }
         } else {
@@ -1742,8 +1757,11 @@ bool ap_wlan_hal_nl80211::update_vap_credentials(
                     conf.set_create_vap_value(vap_id, "multi_ap_backhaul_wpa_passphrase", "");
                 }
 
+                // TODO: Disable management frames for backhaul ap vap to prevent STA disconnection
                 // we always need to get the mgmt frames (assoc req) for capability reports:
-                conf.set_create_vap_value(vap_id, "notify_mgmt_frames", "1");
+                if (bss_it.fronthaul) {
+                    conf.set_create_vap_value(vap_id, "notify_mgmt_frames", "1");
+                }
 
                 // finally enable the vap (remove any previously set start_disabled)
                 conf.set_create_vap_value(vap_id, "start_disabled", "");
@@ -3112,7 +3130,7 @@ bool ap_wlan_hal_nl80211::reload_all_ap_vaps()
             LOG(DEBUG) << "Link " << link_id << " operational status: " << (link_operational ? "true" : "false");
             
             if (!link_operational) {
-                LOG(WARNING) << "Link " << link_id << " is not operational, skipping RELOAD";
+                LOG(WARNING) << "Link " << link_id << " is not operational, trigger RELOAD";
                 std::string hostapd_ctrl_path = "/var/run/hostapd/" + m_radio_info.iface_name;
                 if (beerocks::os_utils::file_exists(hostapd_ctrl_path)) {
                     LOG(INFO) << "Sending RELOAD to " << m_radio_info.iface_name;
diff --git a/common/beerocks/hostapd/source/configuration.cpp b/common/beerocks/hostapd/source/configuration.cpp
index 96c11c9e8..684d82026 100644
--- a/common/beerocks/hostapd/source/configuration.cpp
+++ b/common/beerocks/hostapd/source/configuration.cpp
@@ -306,7 +306,7 @@ bool Configuration::set_create_vap_value(const std::string &vap, const std::stri
         m_last_message =
             std::string(__FUNCTION__) + " the key '" + key + "' for vap " + vap + " was not found";
 
-        if (value.length() == 0) {
+        if (value.empty() || value == "0") {
             LOG(DEBUG) << "key is not found and value is empty - no need to update ";
             return true;
         }
-- 
2.34.1

